functional-samples/ai.gemini-on-device/sidepanel/index.js

```diff
@@ -0,0 +1,130 @@
+      const canCreate = await self.ai.canCreateTextSession();
+      // canCreate will be one of the following:
+      // * "readily": the model is available on-device and so creating will happen quickly
+      // * "after-download": the model is not available on-device, but the device is capable,
+      //   so creating the session will start the download process (which can take a while).
+      // * "no": the model is not available for this device.
+      if (canCreate === 'no') {
+        console.warn('Built-in prompt API not available.');
+        throw new Error(
+          'Built-in prompt API not available. Join the preview program to learn how to enable it.'
+        );
+      }
+      console.log('Creating new text session');
+      session = await self.ai.createTextSession(params);
+    }
+    return session.prompt(prompt);
+  } catch (e) {
+    console.log('Prompt failed');
+    console.error(e);
+    console.log('Prompt:', prompt);
+    // Reset session
+    session = null;
+    throw e;
+  }
+}
+
+async function reset() {
+  if (session) {
+    session.destroy();
+  }
+  session = null;
+}
+
+async function initDefaults() {
+  const defaults = await window.ai.defaultTextSessionOptions();
+  console.log('Model default:', defaults);
+  sliderTemperature.value = defaults.temperature;
+  sliderTopK.value = defaults.topK;
+  labelTopK.textContent = defaults.topK;
+  labelTemperature.textContent = defaults.temperature;
+  labelTemperature.value = defaults.temperature;
+}
+
+initDefaults();
+
+buttonReset.addEventListener('click', () => {
+  hide(elementLoading);
+  hide(elementError);
+  hide(elementResponse);
+  reset();
+  buttonReset.setAttribute('disabled', '');
+});
+
+sliderTemperature.addEventListener('click', reset.bind.this);
+sliderTemperature.addEventListener('input', (event) => {
+  labelTemperature.textContent = event.target.value;
+});
+
+sliderTopK.addEventListener('click', reset.bind.this);
+sliderTopK.addEventListener('input', (event) => {
+  labelTopK.textContent = event.target.value;
+});
+
+inputPrompt.addEventListener('input', () => {
+  if (inputPrompt.value.trim()) {
+    buttonPrompt.removeAttribute('disabled');
+  } else {
+    buttonPrompt.setAttribute('disabled', '');
+  }
+});
+
+buttonPrompt.addEventListener('click', async () => {
+  const prompt = inputPrompt.value.trim();
+  showLoading();
+  try {
+    const params = {
+      temperature: sliderTemperature.value,
+      topK: sliderTopK.value
+    };
+    const response = await runPrompt(prompt, params);
+    showResponse(response);
+  } catch (e) {
+    showError(e);
+  }
+});
+
+function showLoading() {
+  buttonReset.removeAttribute('disabled');
+  hide(elementResponse);
+  hide(elementError);
+  show(elementLoading);
+}
+
+function showResponse(response) {
+  hide(elementLoading);
+  show(elementResponse);
+  elementResponse.textContent = response;
```

Comment:
Something about this seems to lose newlines from the response. Can we see what is needed to preserve those?

Reply:
Good point, replacing \n with <br>. Ideally, I was considering using a markdown renderer here (as gemini really likes to format responses in markdown), but I'd prefer to keep the formatting here.

Reply:
Hmm, can we find an alternative to using innerHTML? Maybe splitting by newline and creating a separate child div per line? With the current code it would render any returned code, which feels bad. We could do some manual escaping but it seems safer just to avoid innerHTML entirely. Also happy to leave for a follow-up, preferably by reverting to the original code for this PR.

Reply:
You're right. Doing the right thing now...

Reply:
Looks great!